formats/hocon/src/main/kotlin/kotlinx/serialization/hocon/Hocon.kt

```diff
@@ -141,7 +141,9 @@ public sealed class Hocon(
             when {
                 descriptor.kind.listLike -> ListConfigReader(conf.getList(currentTag))
                 descriptor.kind.objLike -> if (ind > -1) ConfigReader(conf.getConfig(currentTag)) else this
-                descriptor.kind == StructureKind.MAP -> MapConfigReader(conf.getObject(currentTag))
+                descriptor.kind == StructureKind.MAP ->
+                    //if current tag is null - map in the root of config
+                    MapConfigReader(if (currentTagOrNull != null) conf.getObject(currentTag) else conf.root())
```

Comment:
This fix indeed fixes reported problem, but while we are here, I'd suggest to fix all similar problems consistently and thoughtfully.

What else here worth investigating:
- Top-level lists. If they are not supported in HOCON, then clear exception message
- Top-level primitives. Ditto as with lists
- Nullability. Test `Map<..>?` both when it's null (if it's supported) and when it's not. Same for lists/primitives

Reply:
1. root-level lists supported in HOCON spec https://github.com/lightbend/config/blob/master/HOCON.md#syntax but typesafe config library support only subset of HOCON syntax https://github.com/lightbend/config/issues/460. So library not decode root-level lists
2. root-level primitives are not supported by HOCON and typesafe lib
3. root-level Map<>? supported but have unexpected behavior: if decode empty object "{}" not-null map reads as empty and nullable map reads as null. There is no way to define "null" config because primitives are illegal